hexstody-hot/src/main.rs

```diff
@@ -35,41 +38,118 @@ struct Args {
 #[derive(Parser, Debug, Clone)]
 enum SubCommand {
     /// Start listening incoming API requests
-    Serve
+    Serve,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+enum ApiType {
+    Public,
+    Operator,
+}
+
+impl fmt::Display for ApiType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ApiType::Public => write!(f, "Public"),
+            ApiType::Operator => write!(f, "Operator"),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct ApiConfig {
+    public_api_enabled: bool,
+    public_api_port: u16,
+    operator_api_enabled: bool,
+    operator_api_port: u16,
+}
+
+fn parse_api_config() -> ApiConfig {
+    let figment = rocket::Config::figment();
+    let public_api_enabled = figment.extract_inner("public_api_enabled").unwrap_or(true);
+    let public_api_port = figment.extract_inner("public_api_port").unwrap_or(8000);
+    let operator_api_enabled = figment
+        .extract_inner("operator_api_enabled")
+        .unwrap_or(true);
+    let operator_api_port = figment.extract_inner("operator_api_port").unwrap_or(8001);
+    ApiConfig {
+        public_api_enabled,
+        public_api_port,
+        operator_api_enabled,
+        operator_api_port,
+    }
+}
+
+async fn serve_api(
+    pool: Pool,
+    state_mx: Arc<Mutex<State>>,
+    state_notify: Arc<Notify>,
+    api_type: ApiType,
+    api_enabled: bool,
+    port: u16,
+) -> () {
+    if !api_enabled {
+        info!("{api_type} API disabled");
+        return ();
+    };
+    info!("Serving {api_type} API");
+    match api_type {
+        ApiType::Public => loop {
+            let (_abort_api_handle, abort_api_reg) = AbortHandle::new_pair();
+            let api_fut = tokio::spawn(serve_public_api(
+                pool.clone(),
+                state_mx.clone(),
+                state_notify.clone(),
+                port,
+            ));
+            match Abortable::new(api_fut, abort_api_reg).await {
```

Comment:
api_fut is JoinHandle that cannot be aborted that way. You need to move termination logic into tokio::spawn body and terminate inner future on abortion.